File: backend/app/service/torneo_service.py
# =============================================================================
# SERVICE TORNEO - intermediario tra controller e repository per i tornei.
# =============================================================================

from app.model.torneo import Torneo                              # modello torneo
from app.repository.torneo_repository import TorneoRepository    # accesso al database

class TorneoService:

    def __init__(self):
        self.torneo_repository = TorneoRepository()  # crea il collegamento al repository

    def get_tornei(self) -> list[Torneo]:
        """Restituisce tutti i tornei."""
        return self.torneo_repository.get_tornei()

    def get_torneo_by_id(self, id: int) -> Torneo:
        """Restituisce un torneo dato il suo id."""
        return self.torneo_repository.get_torneo_by_id(id)

    def create_torneo(self, torneo: Torneo) -> int:
        """Crea un nuovo torneo e ritorna l'id assegnato."""
        return self.torneo_repository.create_torneo(torneo)

    def update_torneo(self, torneo: Torneo) -> bool:
        """Aggiorna un torneo esistente; True se trovato e modificato."""
        return self.torneo_repository.update_torneo(torneo)

    def delete_torneo(self, id: int) -> bool:
        """Elimina un torneo; True se trovato e cancellato."""
        return self.torneo_repository.delete_torneo(id)

File: frontend/src/services/torneoService.js
// =============================================================================
// torneoService.js - funzioni per parlare col backend riguardo ai tornei.
// =============================================================================

import { apiRequest } from './httpClient'

const root = '/torneo'

export const torneoService = {
  // Elenco di tutti i tornei (pubblico).
  getTornei: () => apiRequest(`${root}/`),

  // Dettaglio di un torneo dato il suo id.
  getTorneoById: (id) => apiRequest(`${root}/${id}`),

  // Crea un nuovo torneo. Serve il token: il backend usa il JWT per
  // associare il torneo al suo proprietario (l'utente loggato).
  createTorneo: (payload, token) =>
    apiRequest(`${root}/`, { method: 'POST', body: payload, auth: true, token }),

  // Aggiorna un torneo (solo il proprietario).
  updateTorneo: (id, payload, token) =>
    apiRequest(`${root}/${id}`, { method: 'PUT', body: payload, auth: true, token }),

  // Cancella un torneo (solo il proprietario).
  deleteTorneo: (id, token) =>
    apiRequest(`${root}/${id}`, { method: 'DELETE', auth: true, token }),
}
